apisix/plugins/ai-request-rewrite.lua

```diff
@@ -0,0 +1,212 @@
+    description = "Key/value settings for the model",
+    type = "object",
+    properties = {
+        model = {
+            type = "string",
+            description = "Model to execute."
+        }
+    },
+    additionalProperties = true
+}
+
+local schema = {
+    type = "object",
+    properties = {
+        prompt = {
+            type = "string",
+            description = "The prompt to rewrite client request."
+        },
+        provider = {
+            type = "string",
+            description = "Name of the AI service provider.",
+            enum = {"openai", "openai-compatible", "deepseek"} -- add more providers later
+        },
+        auth = auth_schema,
+        options = model_options_schema,
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            maximum = 60000,
+            default = 30000,
+            description = "timeout in milliseconds"
+        },
+        keepalive = {
+            type = "boolean",
+            default = true
+        },
+        keepalive_pool = {
+            type = "integer",
+            minimum = 1,
+            default = 30
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        },
+        override = {
+            type = "object",
+            properties = {
+                endpoint = {
+                    type = "string",
+                    description = "To be specified to override " ..
+                    "the endpoint of the AI service provider."
+                }
+            }
+        }
+    },
+    required = {"prompt", "provider", "auth"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1073,
+    name = plugin_name,
+    schema = schema
+}
+
+local function proxy_request_to_llm(conf, request_table, ctx)
+    local ai_driver = require("apisix.plugins.ai-drivers." .. conf.provider)
+
+    local extra_opts = {
+        endpoint = core.table.try_read_attr(conf, "override", "endpoint"),
+        query_params = conf.auth.query or {},
+        headers = (conf.auth.header or {}),
+        model_options = conf.options
+    }
+
+    local res, err, httpc = ai_driver:request(conf, request_table, extra_opts)
+
+    if not res then return nil, err, nil end
+    return res, nil, httpc
+end
+
+
+local function parse_llm_response(res_body)
+    local response_table = core.json.decode(res_body)
+
+    if not response_table then return nil, "failed to decode llm response" end
+
+    if not response_table.choices then
+        return nil, "'choices' not in llm response"
+    end
+
+    return response_table.choices[1].message.content, nil
+end
+
+
+function _M.check_schema(conf)
+    -- openai-compatible should be used with override.endpoint
+    if conf.provider == "openai-compatible" then
+        local override = conf.override
+
+        if not override or not override.endpoint then
+            return false, "override.endpoint is required for openai-compatible provider"
+        end
+    end
+
```

Comment:
can json-schema do this job?

Reply:
> can json-schema do this job?

As far as I know, JSON Schema is not very good at handling logical checks.

Reply:
@LiteSun json schema can do this check, ref:
https://github.com/apache/apisix/blob/52d8fea3dbc93c4b59ac6682902955c6146af4ad/apisix/schema_def.lua#L836-L849

Reply:
> @LiteSun json schema can do this check, ref:

It works but error message is `"failed to check the configuration of plugin ai-request-rewrite err: then clause did not match"`, it's not a friendly error message. I think current is better.